nsg-netbox.py

```diff
@@ -50,87 +90,311 @@ def run(self):
+            while True:
+                tasks = self.nsg.get_tasks()
+                if not tasks:
+                    break
+                time.sleep(2)
+
+        try:
+            tasks = self.nsg.get_tasks()
             if tasks:
                 self.log.info('NetSpyGlass: tasks: {0}'.format(len(tasks)))
                 self.log.warning('skipping cycle because of an active NSG background task')
+                # schedule next run
+                self.scheduler.enter(delay=self.interval_sec, priority=1, action=self.run)
                 return
 
-            nbox = pynetbox.api(url=self.args.netbox_url, token=self.args.netbox_token)
+            nbox_devices = self.get_netbox_devices()
+            self.log.info(f'Netbox:      {len(nbox_devices):>6} devices')
 
-            nbox_devices = self.get_netbox_devices(nbox)
-            self.log.info('Netbox:      {0} devices'.format(len(nbox_devices)))
+            nsg_devices = self.nsg.get_devices()
+            nsg_devices = {k: v for k, v in nsg_devices.items()
+                           if v.get('address') not in (self.config.get('nsg_blacklist').get('address', [])
+                                                       if self.config.get('nsg_blacklist') else [])
+                           and v.get('name') not in (self.config.get('nsg_blacklist').get('name', [])
+                                                     if self.config.get('nsg_blacklist') else [])
+                           and v.get('id') not in (self.config.get('nsg_blacklist', {}).get('deviceId', [])
+                                                   if self.config.get('nsg_blacklist') else [])
+                           }
+            self.log.info(f'NetSpyGlass: {len(nsg_devices):>6} devices')
 
-            nsg_devices = nsg.get_devices()
-            self.log.info('NetSpyGlass: {0} devices'.format(len(nsg_devices)))
+            tags_ = self.nsg.get_device_tags()
+            tags = {}
+            for i in tags_:
+                key = i.get("device", {}).get("address")
+                if not tags.get(key):
+                    tags[key] = {"id": i.get("device", {}).get("id"),
+                                 "tags": dict(t.split(".", maxsplit=1) for t in i.get("tags", {}))}
+                else:
+                    tags[key]["tags"].update(dict(t.split(".", maxsplit=1) for t in i.get("tags", {})))
 
             nb_dev_set = set(nbox_devices.keys())
             nsg_dev_set = set(nsg_devices.keys())
             to_add = set.difference(nb_dev_set, nsg_dev_set)  # set of addresses as strings
             to_remove = set.difference(nsg_dev_set, nb_dev_set)  # set of addresses as strings
 
+            if to_remove:
+                self.log.info('DELETE devices: {0}'.format(to_remove))
+                self.nsg.delete_devices(list(nsg_devices[addr]['id'] for addr in to_remove))
+                wait_tasks_done()
+                remove_tags = make_remove_tag_dict(nsg_devices=to_remove, nsg_tags=tags)
```

Comment:
in the code below, you add tags if new tags appear in NetBox even when you do not add any new devices. However you remove tags only when you remove devices. What happens if a tag is removed in NetBox but the device is not removed?

Reply:
added tag comparing for all devices